### DISCUSSION TITLE
[@types/aws-lambda] [@aws-sdk/client-dynamdb] AttributeValue definition is different in client dbstream

### DISCUSSION THREAD

**Discussion by ptitkosmos** [DefinitelyTyped/DefinitelyTyped#67358] [Issues with a @types package]
Hello,

the definition of AttributeValue is different between the client library and the type library : 
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4809c86f6f74f88e01713ec69c9e307776c8a299/types/aws-lambda/trigger/dynamodb-stream.d.ts#L7
https://github.com/aws/aws-sdk-js-v3/blob/284bd004ee7c1892ce542e578996b6a53434a721/clients/client-dynamodb-streams/src/models/models_0.ts#L652

Is it really not the same type ? If not, the impact is that the compilator rejects when wanting to unmarshall data.
Thanks

[a comment by an automated account is not shown]

**Comment by jamesorlakin** [comment]
They are different types but conceptually represent the same thing. DefinitelyTyped's definition is based upon all the possible entries that could be returned in a document's record, whereas the AWS SDK is a bit more complex (using type unions) to try and mutually exclude them - e.g. a value with `N: ...` can't also have `S: ...`, it'd have to be one or the other. Stricly speaking that's 'more correct' in practice but it can be harder to follow. Most AWS SDKs are auto-generated so I suspect there's not much maintenance overhead on their side.

> the impact is that the compilator rejects when wanting to unmarshall data.

Is there a specific issue you're running into? In case you didn't know, in TypeScript you don't typically unmarshal into specific classes or structs like you would in Java or Go. JSON gets parsed into an object and you can lookup any key at runtime. The type definitions are useful to guide and check you're interacting with a data structure correctly but they have no meaning at runtime.

The DefinitelyTyped package is aimed at writing Lambda functions in TypeScript that react to an event, e.g. from a DynamoDB stream calling your TypeScript Lambda function directly as a trigger. The AWS SDK is aimed at writing an application that will sit running for long periods consuming the stream itself - it wouldn't be the right fit for Lambda and its 15 minute timeout. Conceptually I think the choice is whether you'd like to use a Lambda function or not and that should guide you as to which package (and general architecture) that'll work for your needs. 🙂

**Comment by ptitkosmos** [comment]
The use case is having a handler triggered by a dynamoDbStream and the will is to deal with objects instead of dbRecords.
```
import { Handler, DynamoDBStreamEvent } from 'aws-lambda';
import { unmarshall } from '@aws-sdk/util-dynamodb';
export const handler: Handler = async (event: DynamoDBStreamEvent) => {
   for (const record of event.Records) {
        if (record.dynamodb) {
              // unmarshall ???
              // const myDbObject = AWS.DynamoDB.Converter.unmarshall(record.dynamodb.OldImage); // aws-sdk V2
              const myDbObject = unmarshall(record.dynamodb.OldImage); // V3 but type mismatch
        }
   }
}
```
The error raised is : 
```
Argument of type '{ [key: string]: AttributeValue; }' is not assignable to parameter of type 'AttributeValue | Record<string, AttributeValue>'.
  Property '$unknown' is missing in type '{ [key: string]: AttributeValue; }' but required in type '$UnknownMember'.ts(2345)
models_0.d.ts(6231, 9): '$unknown' is declared here.
```
Maybe i'm going wrong. How can I handle such case ?

**Comment by klauszhang** [comment]
```ts
import { AttributeValue } from '@aws-sdk/client-dynamodb'
```

This  fixed the issue for me

The types from @types/aws-lambda and @aws-sdk/client-dynamodb for `AttributeValue` are different

**Comment by JohannesKohlmann** [comment]
We have also been running into this issue in a use-case similar to @ptitkosmos. Is there any solution to this that does not require casting the `NewImage` to `Record<string, AttributeValue>`?

**Comment by jimmyn** [comment]
Hello, is there any update on this? This is still an issue.
